### DISCUSSION TITLE
26MB generated header file?

### DISCUSSION THREAD

**Discussion by DUOLabs333** [capnproto/capnproto#1918] [Q&A]
I'm working on moving my program from JSON to Capn Proto. However, I'm autogenerating a lot of bindings, so the resulting capnp file will naturally be large --- around 450KB. However, when I went to compile the file, the header blew up to 26MB. I tried precompiling the header to save compilation time, but I had to kill the process when the `gch` file was reaching 1GB.

**Comment by kentonv** [comment]
Yes, the generated code is much larger than the input schema, since each field declaration turns into several accessor methods, etc.

Why are you auto-generating your schema? This sounds like an unusual way to use Cap'n Proto.

**Reply by DUOLabs333** [reply]
I'm serializing and deserializing structs back and forth between two different systems.

**Reply by kentonv** [reply]
I think absolutely every user of Cap'n Proto could describe their usage that way. :)

That doesn't answer: Why are you auto-generating schemas? Are you translating them from a different format or something? Why is the schema so huge? Can you tell me about it?

**Reply by DUOLabs333** [reply]
I'm generating bindings from Vulkan's XML registry (I'm streaming Vulkan commands). There are a lot of structs and commands that need serde (~2000).

**Reply by kentonv** [reply]
Do you actually expect to write code that manipulates these commands as Cap'n Proto messages, or are you intending to use this schema only to guide translation to/from e.g. JSON or other formats? If the latter, you may not need to generate code in the first place. E.g. the JsonCodec API does not require that the schemas you give it are actually complied into the binary. You can instead load them dynamically from `schema.capnp` format (which you can ask the `capnp` tool to output for you, instead of generating code).

If you are planning to write C++ code that directly manipulates the Cap'n Proto format for these commands, you might consider pruning the schema down to just the commands you actually expect to use.

**Reply by DUOLabs333** [reply]
Cap'n Proto is only at the messaging stage --- once I receive the message, it's immediately being converted back into the arguments.

Even if I only create schemas for commands I can expect to use, I'm still looking at 22MB of code.
